skaffold-plugins-core/src/main/java/com/google/cloud/tools/skaffold/downloader/SkaffoldDownloader.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.skaffold.downloader;
+
+import com.google.cloud.tools.skaffold.filesystem.OperatingSystem;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.HashSet;
+import java.util.Set;
+
+/** Downloads {@code skaffold} executable. */
+public class SkaffoldDownloader {
+
+  /**
+   * Downloads the latest {@code skaffold} release.
+   *
+   * @throws IOException if an I/O exception occurs during download
+   * @throws MalformedURLException if the URL to download from is malformed
+   */
+  public static void downloadLatest(Path destination) throws IOException {
+    download(new URL(getUrl("latest", OperatingSystem.resolve())), destination);
+  }
+
+  /**
+   * Downloads to the {@code destination}.
+   *
+   * @param destination the destination file to download {@code skaffold} to
+   * @throws IOException if an I/O exception occurs during download
+   */
+  @VisibleForTesting
+  static void download(URL url, Path destination) throws IOException {
+    if (Downloader.download(url, destination) == -1) {
```

Comment:
Would it make more sense for this to throw an exception instead of returning -1?

Reply:
I'm trying to be lighter in terms of using exceptions in this new codebase. If this was changed to an `IOException`, we wouldn't be able to differentiate having a non-positive length from `IOException`s that occurred (which could be all sorts of problems). This way, it also gives us the ability to choose what to do upon no content length, which here is to say that specifically, the size of the `skaffold` binary could not be resolved.

Reply:
It just seems awkward to have something throw exceptions for a bunch of different error cases and then return -1 for a different error case (especially when all we're doing in that case is throwing another exception). As far as differentiating goes we could always make our own exception or do an error message comparison. Although I understand the desire to be lighter with exceptions, and maybe this is just the neatest way to do it. Just stuff to consider though.